src/tests/test_sexual_behaviour.py

```diff
@@ -104,3 +106,225 @@ def test_initial_sex_behaviour_groups():
+    # need to set the number of new partners to test for risk reduction cases
+    pop.data["num_partners"] = 0
+    men = selector(pop.data, sex=(operator.eq, SexType.Male))
+    women = selector(pop.data, sex=(operator.eq, SexType.Female))
+
+    # equal numbers of partners
+    n_partners_men = n_partners//2
+    n_partners_women = n_partners//2
+    # distribute partners amongst the men and women
+    men_idx = pop.data.loc[men].index
+    women_idx = pop.data.loc[women].index
+    for i in range(n_partners_men):
+        pop.data.loc[rng.choice(men_idx), "num_partners"] += 1
+    for i in range(n_partners_women):
+        pop.data.loc[rng.choice(women_idx), "num_partners"] += 1
+    SBM.update_rred_balance(pop.data)
+    assert np.all(pop.data.loc[men, "rred_balance"] == 1)
+    assert np.all(pop.data.loc[women, "rred_balance"] == 1)
+
+    # ~ -1 % discrepancy
+    pop.data["num_partners"] = 0
+    n_partners_men = int(n_partners * 0.495)
+    n_partners_women = int(n_partners * 0.505)
+    for i in range(n_partners_men):
+        pop.data.loc[rng.choice(men_idx), "num_partners"] += 1
+    for i in range(n_partners_women):
+        pop.data.loc[rng.choice(women_idx), "num_partners"] += 1
+    SBM.update_rred_balance(pop.data)
+    assert np.allclose(pop.data.loc[men, "rred_balance"], 1/0.7)
+    assert np.allclose(pop.data.loc[women, "rred_balance"], 0.7)
+
+    # > 10 % discrepancy
+    pop.data["num_partners"] = 0
+    n_partners_men = int(n_partners * 0.6)
+    n_partners_women = int(n_partners * 0.45)
+    for i in range(n_partners_men):
+        pop.data.loc[rng.choice(men_idx), "num_partners"] += 1
+    for i in range(n_partners_women):
+        pop.data.loc[rng.choice(women_idx), "num_partners"] += 1
+    SBM.update_rred_balance(pop.data)
+    assert np.all(pop.data.loc[men, "rred_balance"] == 0.1)
+    assert np.all(pop.data.loc[women, "rred_balance"] == 10)
+
+
+def test_rred_art_adherence():
+    N = 1000
+    pop = Population(size=N, start_date=date(1989, 1, 1))
+    SBM = SexualBehaviourModule()
+    SBM.init_risk_factors(pop.data)
+    # art adherence flag off
+    SBM.rred_art_adherence_flag = False
+    pop.data["art_adherence"] = rng.random(size=N)
+    above_idx = selector(pop.data, art_adherence=(operator.ge, 0.8))
+    below_idx = selector(pop.data, art_adherence=(operator.lt, 0.8))
+    SBM.update_rred(pop)
+    for i in range(N):
+        if (pop.data["rred_art_adherence"][i] != 1):
+            print(i, pop.data["rred_art_adherence"][i])
+    assert np.allclose(pop.data["rred_art_adherence"], 1)
+
+    # art adherence flag on
+    SBM.rred_art_adherence_flag = True
+    SBM.update_rred(pop)
+    assert np.all(pop.data.loc[above_idx, "rred_art_adherence"] == 1)
+    assert np.all(pop.data.loc[below_idx, "rred_art_adherence"] == 2)
+
+    count = 0
+    for i in range(100):
+        SBM_temp = SexualBehaviourModule()
+        count += SBM_temp.rred_art_adherence_flag
+    assert 0.1 < count/100 < 0.3
```

Comment:
I feel this should be its own test as it's checking something very different from the parts above.